File: services/auth-service/internal/middleware/ratelimit.go
package middleware

import (
	"net/http"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
)

// slidingWindow tracks request timestamps per client key.
type slidingWindow struct {
	mu      sync.Mutex
	hits    map[string][]time.Time
	limit   int
	window  time.Duration
	lastGC  time.Time
	gcEvery time.Duration
}

func newSlidingWindow(limit int, window time.Duration) *slidingWindow {
	return &slidingWindow{
		hits:    make(map[string][]time.Time),
		limit:   limit,
		window:  window,
		lastGC:  time.Now(),
		gcEvery: 5 * time.Minute,
	}
}

// allow reports whether the key may make another request now.
func (s *slidingWindow) allow(key string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()

	now := time.Now()
	cutoff := now.Add(-s.window)

	// Periodic GC so idle keys do not leak memory.
	if now.Sub(s.lastGC) > s.gcEvery {
		for k, ts := range s.hits {
			if len(ts) == 0 || ts[len(ts)-1].Before(cutoff) {
				delete(s.hits, k)
			}
		}
		s.lastGC = now
	}

	ts := s.hits[key]
	kept := ts[:0]
	for _, t := range ts {
		if t.After(cutoff) {
			kept = append(kept, t)
		}
	}

	if len(kept) >= s.limit {
		s.hits[key] = kept
		return false
	}

	s.hits[key] = append(kept, now)
	return true
}

// RateLimit returns a per-IP sliding-window rate limiter middleware.
// Intended for sensitive endpoints (login, register, refresh).
// In-memory by design: swap for the Redis sliding window at the API gateway
// when running multiple replicas.
func RateLimit(limit int, window time.Duration) gin.HandlerFunc {
	sw := newSlidingWindow(limit, window)
	return func(c *gin.Context) {
		if !sw.allow(c.ClientIP()) {
			c.AbortWithStatusJSON(http.StatusTooManyRequests, gin.H{
				"success": false,
				"error": gin.H{
					"code":    "RATE_LIMITED",
					"message": "Bạn thao tác quá nhanh. Vui lòng thử lại sau ít phút.",
				},
			})
			return
		}
		c.Next()
	}
}

File: services/ai-service/app/core/ratelimit.py
"""In-memory sliding-window rate limiter middleware.

AI endpoints are expensive (LLM calls), so each client IP gets a small
per-minute budget. Single-process by design — replace with the Redis
sliding window at the API gateway when running multiple replicas.
"""

import time
from collections import defaultdict

from fastapi import Request
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.types import ASGIApp

_EXEMPT_PATHS = ("/health", "/docs", "/openapi.json")


class RateLimitMiddleware(BaseHTTPMiddleware):
    """Limit each client IP to `limit` requests per `window_seconds`."""

    def __init__(self, app: ASGIApp, limit: int = 20, window_seconds: int = 60) -> None:
        super().__init__(app)
        self.limit = limit
        self.window = window_seconds
        self._hits: dict[str, list[float]] = defaultdict(list)
        self._last_gc = time.monotonic()

    def _allow(self, key: str) -> bool:
        now = time.monotonic()
        cutoff = now - self.window

        # Periodic GC so idle keys don't accumulate.
        if now - self._last_gc > 300:
            for k in [k for k, ts in self._hits.items() if not ts or ts[-1] < cutoff]:
                del self._hits[k]
            self._last_gc = now

        hits = [t for t in self._hits[key] if t > cutoff]
        if len(hits) >= self.limit:
            self._hits[key] = hits
            return False

        hits.append(now)
        self._hits[key] = hits
        return True

    async def dispatch(self, request: Request, call_next):  # type: ignore[no-untyped-def]
        if request.url.path in _EXEMPT_PATHS:
            return await call_next(request)

        client_ip = request.client.host if request.client else "unknown"
        if not self._allow(client_ip):
            return JSONResponse(
                status_code=429,
                content={
                    "success": False,
                    "error": {
                        "code": "RATE_LIMITED",
                        "message": "Bạn thao tác quá nhanh. Vui lòng thử lại sau ít phút.",
                    },
                },
            )
        return await call_next(request)
